### DISCUSSION TITLE
Save a signed version

### DISCUSSION THREAD

**Discussion by mariovalney** [MatthiasValvekens/pyHanko#211] [General]
(I'm not sure it's already possible.)

Would be great to be able to manage / write a early signed version of document.

Just like Adobe Acrobat:

![image](https://user-images.githubusercontent.com/5784633/215881185-e8f3dd59-a15c-46e4-bb82-8c98d2c8f862.png)

I've tried to write a "HistoricalResolver" without success.

**Comment by MatthiasValvekens** [comment]
Hi @mariovalney,

I'm not sure what you mean. PDF signatures are always applied in this manner: the first signature applies to the "base" revision, and further signatures are added incrementally as "later" revisions.

Of course, if you want to sign a revision and then make changes to the _document contents_ (read: anything other than adding more signatures or messing with form data/annotations), then a serious validator will invalidate all preceding signatures. Acrobat does the same, for the record.

Either way, I'm going to need some clarification as to what your use case is :).

FWIW: ``HistoricalResolver`` is a forensic tool meant for reading files, not writing them. If making incremental updates (=> new revisions) is what you want to do, then have a look at `IncrementalPdfFileWriter`.

**Comment by mariovalney** [comment]
Hey! Thanks for your answer.
Sorry for not being very descriptive.

I'm intending to "export" or "save" a signed version of PDF.

Use case:

1. A PDF is signed in ICP-Brasil standard by a user.
2. Docusign plataform add another sign with its certificate which is not valid in ICP-Brasil.
3. We have to open PDF in Adobe and open and save the revision before Docusign's signature.

**Reply by MatthiasValvekens** [reply]
This is more complicated than it sounds, depending on how much info you want to preserve. If you are OK with destroying _all_ prior signatures in the document, this is how I'd do it:

 - Open the document with `PdfFileReader` and load it into a fresh `PdfFileWriter` using `copy_into_new_writer(...)`.
 - Delete the existing signature field(s) in the new `PdfFileWriter`.
 - Use that same `PdfFileWriter` to produce a new signature that is to your liking.

Note that this will cause the entire document to be reserialised, which may be too invasive, especially if you care about any prior non-Docusign signatures.

If you do need to preserve more of the document, then there's no universal way to proceed that will work in all scenarios.

 - Using `HistoricalResolver` allows you to read the revision prior to the one containing the signature, but it doesn't tell you exactly where the `%%EOF` for that revision is, so cutting off the file at the right point is tricky.
 - However, you can use `reader.xref_cache.get_xref_container_info(revision).end_location` to find the end of the xref data of a revision, and read forward until you encounter the `%%EOF` marker. Not pretty, but there's a chance that it works.
 - Alternatively, you could find the **ByteRange** of the existing signature and try to put your signature into the "hole" carved out for Docusign's signature container. There's a low level API in pyHanko that will allow you to do that (essentially by short-circuiting the signing process until the very last step), but it requires your new signature to actually fit in that space. That may or may not be realistic.

What do you think?

**Reply by mariovalney** [reply]
> If you are OK with destroying all prior signatures in the document, this is how I'd do it.

It's the quite opposite. I need all prior signatures so I'm OK destroying later signatures (Docusign).



> What do you think?

I'll give a try in the other options... maybe more tests.
You gave me a way to understand better how to use `reader.xref_cache.get_xref_container_info(revision).end_location`. 
Thanks! I'll let you know about my progress.
